src/fv/share/tdyugrid.c

```diff
@@ -0,0 +1,1209 @@
+///
+/// @param [in] ugrid A TDyUGrid struct
+/// @param [inout] AdjMat Adjacency matrix
+///
+/// @returns 0 on success, or a non-zero error code on failure
+static PetscErrorCode CreateAdjacencyMatrix(TDyUGrid *ugrid, Mat *AdjMat) {
+
+  PetscErrorCode ierr;
+
+  // Determine the i-th and j-th values for creating the adjacency matrix:
+  //   - 	i: the indices into j for the start of each row
+  //   - 	j: the column indices for each row (sorted for each row).
+  PetscInt *i, *j;
+  PetscInt nrow = ugrid->num_cells_local;
+  PetscInt ncol = ugrid->max_verts_per_cell;
+
+  ierr = TDyAllocate_IntegerArray_1D(&i, nrow + 1); CHKERRQ(ierr);
+  ierr = TDyAllocate_IntegerArray_1D(&j, nrow*ncol); CHKERRQ(ierr);
+
+  i[0] = 0;
+  PetscInt count=0;
+  for (PetscInt icell=0; icell<nrow; icell++) {
+    for (PetscInt ivert=0; ivert<ncol; ivert++) {
+      if (ugrid->cell_vertices[icell][ivert] >= 0 ){
+      j[count++] = ugrid->cell_vertices[icell][ivert];
+      }
+    }
+    i[icell+1] = count;
+  }
+
+  // Determine the global offset for rows on each rank
+  PetscInt global_offset = 0;
+  ierr = MPI_Exscan(&nrow, &global_offset, 1, MPI_INTEGER, MPI_SUM, PETSC_COMM_WORLD); CHKERRQ(ierr);
+
+  ierr = MatCreateMPIAdj(PETSC_COMM_WORLD, nrow, ugrid->num_verts_global, i, j, PETSC_NULL, AdjMat); CHKERRQ(ierr);
+#ifdef UGRID_DEBUG
+  ierr = TDySavePetscMatAsASCII(*AdjMat, "Adj.out"); CHKERRQ(ierr);
+#endif
+
+  PetscFunctionReturn(0);
+}
+
+/// Use the dual matrix to partition the grid
+///
+/// @param [in] DualMat A dual matrix of the adjaceny matrix
+/// @param [out] NewCellRankIS IS that identifies the rank that owns the grid cell
+/// @param [out] NewNumCellsLocal Number of local cells after partitioning
+///
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode PartitionGrid(Mat DualMat, IS *NewCellRankIS, PetscInt *NewNumCellsLocal) {
+
+   PetscErrorCode ierr;
+
+  // Create partitioning object based on the dual matrix
+  MatPartitioning Part;
+  ierr = MatPartitioningCreate(PETSC_COMM_WORLD, &Part); CHKERRQ(ierr);
+  ierr = MatPartitioningSetAdjacency(Part, DualMat);CHKERRQ(ierr);
+  ierr = MatPartitioningSetFromOptions(Part);CHKERRQ(ierr);
+
+  // Now partition the mesh
+  ierr = MatPartitioningApply(Part, NewCellRankIS); CHKERRQ(ierr);
+  ierr = MatPartitioningDestroy(&Part); CHKERRQ(ierr);
+
+  // Compute the number of local grid cells on each processor
+  PetscInt commsize, myrank;
+  ierr = MPI_Comm_size(PETSC_COMM_WORLD, &commsize); CHKERRQ(ierr);
+  ierr = MPI_Comm_rank(PETSC_COMM_WORLD, &myrank); CHKERRQ(ierr);
+
+  PetscInt cell_counts[commsize];
+  ierr = ISPartitioningCount(*NewCellRankIS, commsize, cell_counts); CHKERRQ(ierr);
+  *NewNumCellsLocal = cell_counts[myrank];
+
+#ifdef UGRID_DEBUG
+  ierr = TDySavePetscISAsASCII(*NewCellRankIS,"is.out");
+#endif
+
+  PetscFunctionReturn(0);
+}
+
+/// Determine the max number of duals/neighbors of a grid cell in the mesh
```

Comment:
This determines the max number of duals/neighbors for all cells in the mesh, not just one.